client/state/receipts/assembler.ts

```diff
@@ -0,0 +1,80 @@
+import { flatten } from 'lodash';
+import type {
+	RawFailedReceiptPurchase,
+	RawFailedReceiptPurchases,
+	RawReceiptData,
+	RawReceiptPurchase,
+	RawReceiptPurchases,
+	ReceiptData,
+} from './types';
+
+/**
+ * Converts raw receipt data into receipt data
+ *
+ * @param {RawReceiptData} data The raw data returned from the server after a transaction
+ * @returns {ReceiptData} The formatted receipt data
+ */
+export function createReceiptObject( data: RawReceiptData ): ReceiptData {
+	const purchases = Array.isArray( data.purchases ) ? {} : data.purchases;
+	const failedPurchases = Array.isArray( data.failed_purchases ) ? {} : data.failed_purchases;
```

Comment:
This is the biggest change I made in this PR. Previously these two lines would provide an empty object only if the purchases/failed_purchases were _falsy_  (eg: `data.purchases || {}`) but according to the type, they are very explicitly either an object or an empty array.

Reply:
I was right here, but also wrong.

If `data.purchases` is falsy, then `Array.isArray(false)` will indeed return false... but then this statement will cause `purchases` to be `false` as well, when it should be `{}`. I've corrected this in https://github.com/Automattic/wp-calypso/pull/69107 by updating the types so it covers the case where `data.purchases` is neither an array nor an object.